File: JAVASCRIPTSDAS/SORTING/04mergeSort.js
function merge(arr, start, mid, end) {
    let Result = [];
    let i = start, j = mid + 1;
  
    while (i <= mid && j <= end) {
      if (arr[i] <= arr[j]) {
        Result.push(arr[i++]);
      } else {
        Result.push(arr[j++]);
      }
    }
  
    while (i <= mid) {
      Result.push(arr[i++]);
    }
  
    while (j <= end) {
      Result.push(arr[j++]);
    }
  
    for (let k = 0; k < Result.length; k++) {
      arr[start + k] = Result[k];
    }
  }
  
  function mergeAlgo(arr, start, end) {
    if (start < end) {
      let mid = Math.floor((start + end) / 2);
      mergeAlgo(arr, start, mid);
      mergeAlgo(arr, mid + 1, end);
      merge(arr, start, mid, end);
    }
  }
  
  function mergeSort(arr) {
    mergeAlgo(arr, 0, arr.length - 1);
    console.log(arr);
  }
  
  let arr = [9, 6, 3, 3, 2, 78, 5, 5, 2];
  mergeSort(arr);
  

File: CPPDSA/SORTING/04mergeSort.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std  ;
void merge(vector<int>&arr , int start , int mid  , int end  ) {
 vector<int> result  ;
 int i = start , j = mid+1 ;

 while ( i <= mid  && j <= end ) {
    if( arr[i] <= arr[j]) {
        result.push_back(arr[i++]) ;
    }else{
        result.push_back(arr[j++]) ;
    }
 }

 while( i <= mid  ) {
     result.push_back(arr[i++]) ;
 }
 while( j <= end ) {
    result.push_back(arr[j++]) ;
 }

for ( int k =0 ; k<result.size() ;  k++ ) {
    arr[start+k] = result[k] ;
}
}

void mergeAlgo( vector<int>&arr , int start  , int end  ) {
  
if ( start >= end   )  return ; 

  int mid = (start + end  )/ 2  ;
  mergeAlgo( arr , start , mid  ) ;
  mergeAlgo(  arr , mid +1 , end  ) ;
  merge( arr , start  , mid , end   ) ;

}
void  mergeSort(vector<int>&arr , int size )  {
    
mergeAlgo( arr , 0 , size-1 ) ;

  
}
int main() {
    vector<int>arr  = { 9,6,6,3,8,5,2,23} ;
    int size  = arr.size () ;
    mergeSort( arr , size ) ;
    cout<<"Sorted array is" ;
    for( int val : arr ){
        cout<< val <<" " ;
    }
    return 0  ;  
}